Un-deprecate multiQuery #984 and pre-fetch properties to cache

smcquillan opened the pull request:
Signed-off-by: Scott McQuillan <scott.mcquillan@uk.ibm.com>

Pull request  to un-deprecates multiQuery for issue #984 

In addition it adds an optimisation to use multiQuery to pre-fetch properties for multiple vertices into the cache if the next step of the query is a `has` step. For queries following large number of edges this eliminates many network trips to the storage backend for the vertex properties and can give significant performance improvements, see the issue for an example.

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

FlorianHockmann commented:
Thanks for the contribution. The avoided networks trips definitely sound nice.
I will try to review this in the next few days.

Travis failed here because of a known flaky test (#1459). I just restarted the job.
Regarding the comments from Codacy: We just activated that app and it definitely generates some false positives. I suggest that you just check its comments to see which are valid complaints so you can fix them and ignore the comments that don't really make sense.

smcquillan commented:
The comments from Codacy were all pretty minor and many were about existing issues. I've just amended the commit to move the constructor to below the field declarations and taken out the unused imports which looks like it'll be enough to keep Codacy happy.

FlorianHockmann commented on janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:
```diff
@@ -4023,6 +4021,8 @@ public void testTinkerPopOptimizationStrategies() {
         assertNumStep(1, 1, gts.V().has("id", numV / 5).has("weight", (numV / 5) % 5).as("x"), JanusGraphStep.class);
 
 
+        assertNumStep(superV * (numV / 5 * 2), 2, gts.V().has("id", sid).outE("knows").has("weight", P.between(1, 3)).inV().has("weight", P.between(1, 3)), JanusGraphStep.class, JanusGraphVertexStep.class);
```
How can these tests verify your changes? I just removed all code changes and only kept these tests here to verify that they fail without your changes but the test method is still passing.
Unfortunately, I don't know how these changes could be tested else. Would it maybe be possible to add tests where a traversal pre-optimization is given and the optimized version is asserted [like it's done in `JanusGraphStepStrategyTest`](https://github.com/JanusGraph/janusgraph/blob/2fb971dd69abcd49e44dea33ccbe0d54a5df5357/janusgraph-test/src/test/java/org/janusgraph/blueprints/process/traversal/strategy/optimization/JanusGraphStepStrategyTest.java#L132)?

smcquillan replied:
I too was unsure how best to test this given that if it's working then it should return exactly the same results but faster. So we'd not expect the tests to fail without the changes, just to be slower when there are a large number of edges.

I verified that the two tests call the new code both by printing out temporary debug to stdout and by placing breakpoints in eclipse and stopping the debugger in the new methods.

I could/should look to add a check for the flavour of optimisation which uses the new  `JanusGraphEdgeVertexStep` class I think the profiler would pick that up. But the profile for the other flavour won't look any different.

smcquillan replied:
There's now a check to ensure that the profiling shows that `multiPreFetch=true` meaning the optimisation did kick in.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/HasStepFolder.java:
```diff
@@ -329,4 +331,26 @@ static void foldInRange(final HasStepFolder janusgraphStep, final Step<?, ?>  ti
         }
     }
+
+	static boolean foldableHasContainer(FlatMapStep<?, ?> janusgraphStep, Admin<?, ?> traversal, Admin<?, ?> traversal2) {
```
Well spotted, I'd used the logic in the `foldInHasContainer` method as a starting point and it had those extra two parameters but they're are not needed in `foldableHasContainer` so I'll remove them.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphLocalQueryOptimizerStrategy.java:
```diff
@@ -169,6 +173,43 @@ private static void unfoldLocalTraversal(final Traversal.Admin<?, ?> traversal,
+		            vertexStep.setTxVertexCacheSize(txVertexCacheSize);
+		        }
+		    }
+		    else if (nextStep instanceof EdgeVertexStep) {
```
There are two different flavours of the optimisation here, hence the condition:

- The first branch handles the `outE().inV().has()` flavour and from within the existing JanusGraphVertexStep class which will, in this case have a return type `Vertex` for `JanusGraphVertexStep.getReturnClass()`

- The second branch handles that case where there is another step like your `outE().has('weight', 0.5).inV().has()` example. And in this case the return type is `Edge` for `JanusGraphVertexStep.getReturnClass()`. It's for this flavour that I added the new `JanusGraphEdgeVertexStep` class which is doing the `.inV().has()` part of the query and loading the properties for those vertices into the cache.

I've not tested what happens if other step is an `as` step like `outE().as('edge').inV()` so will investigate that one further.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphEdgeVertexStep.java:
```diff
@@ -0,0 +1,106 @@
+            starts.forEachRemaining(e -> {
+                edges.add(e);
+
+                if (vertices.size() < txVertexCacheSize) {
```
That's right, and following on from your earlier comment about What happens to a traversal that has a limit() it might be possible to limit it further based on that. I'll investigate.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphEdgeVertexStep.java:
```diff
@@ -0,0 +1,106 @@
+            // populate the vertex cache so subsequent queries of properties don't have to go to the storage back end
+            if (vertices.size() > 1) {
+                JanusGraphMultiVertexQuery multiQuery = JanusGraphTraversalUtil.getTx(traversal).multiQuery();
+                multiQuery.addAllVertices(vertices).properties();
```
This is triggered for queries like `outE().has('weight', 0.5).inV().has()` where return type is Edge for JanusGraphVertexStep.getReturnClass(). In the situation I think there are two calls to `multiQuery` being made. 

- The first in `JanusGraphVertexStep` to get the edge properties and filter on those.
- The second one `JanusGraphEdgeVertexStep` to get the vertex properties after following the edges from the first.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphVertexStep.java:
```diff
@@ -109,7 +120,26 @@ private void initialize() {
             assert multiQueryResults != null;
-            return (Iterator<E>) multiQueryResults.get(traverser.get()).iterator();
+            
+            Iterable<? extends JanusGraphElement> result = multiQueryResults.get(traverser.get());
```
No. This is triggered for the `outE().inV().has()` flavour and from within the existing JanusGraphVertexStep class which will, in this case have a return type Vertex for JanusGraphVertexStep.getReturnClass(). So its skipping right over the edges and returning the vertices at the far end of those edges, it's those vertices we're doing the multiQuery on.

smcquillan commented:
Like you I haven't worked on these parts of the code base before though I have been digging through them with the help the debugger in eclipse for some weeks now. I agree that it would be good if someone familiar with this area of the code could also review this.

Wanted to respond to the comments questions straight away. But I've taken a few actions items such as following up on the use of `limit()` and `as()`. And also looking to see if, at the very least, I can add some tests that use the profiler to verify that the optimisation has been triggered.

smcquillan commented:
@FlorianHockmann I've looked at what happens if we have an intermediate `as` in the query like `outE().as('edge').inV()` and the good news there is that the optimisation does still kick in.

With respect so something like the use of the use of limit in a traversal like the following `g.V().has('name', 'hercules').out().has('age', 5).limit(1)` we've a chicken and egg situation as the limit is applied **after** the has step checks if the age is 5 and the optimisation needs to kick in **before** the has step so not sure there's much we can do there.

Better news on the use of profiling in the tests though as I worked out how to explicitly profile the multiQuery which is fetching the properties and was able to tag it with a new annotation in the query profile, which I've named `_multiPreFetch`. I was able to alter the test to ensure that this annotation was set but better still the profiling data is now much more informative. 

For example this is a query profile from version the  current JanusGraph version `0.3.1`:
```
gremlin> g.V(1050048).outE('aggregation').inV().has('uniqueId', 'interface831').profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[1050048])                                            1           1          37.725     1.14
JanusGraphVertexStep(OUT,[aggregation],vertex)                      1001        1001         158.576     4.78
    \_condition=type[aggregation]
    \_isFitted=true
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                84.298
  backend-query                                                     1001                      20.898
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
HasStep([uniqueId.eq(interface831)])                                   1           1        3123.755    94.09
                                            >TOTAL                     -           -        3320.057        -
```
And here's the same query run with the property pre-fetch optimisation 
```
gremlin> g.V(1050048).outE('aggregation').inV().has('uniqueId', 'interface831').profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[1050048])                                            1           1          37.031     6.60
JanusGraphVertexStep(OUT,[aggregation],vertex)                      1001        1001         404.666    72.17
    \_condition=(PROPERTY AND visibility:normal)
    \_isFitted=true
    \_vertices=1001
    \_multiPreFetch=true
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@8019d62e
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                80.020
  backend-query                                                     1001                      15.789
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
  optimization                                                                                 0.676
  backend-query                                                    13013                     245.383
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@8019d62e
HasStep([uniqueId.eq(interface831)])                                   1           1         119.047    21.23
                                            >TOTAL                     -           -         560.744        -
```
Points to note:
- It's much faster at just over half a second vs. over 3 seconds before. 
- It's tagged with `\_multiPreFetch=true` so we know the optimisation kicked in
- Explicitly shows backend query to get the 13013 properties (1001 vertices with 13 properties each)
- Run a second time and the profile shows it **doesn't** re-fetch, the now cached, 13013 properties

And for reference here's the profiles for the second flavour of query. First for version 0.3.1 
```
gremlin> g.V(1050048).outE('aggregation').has('edgeType', 'contains').inV().has('uniqueId', 'interface831').profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[1050048])                                            1           1          37.682     1.17
JanusGraphVertexStep([edgeType.eq(contains)])                       1001        1001         219.878     6.81
    \_condition=(edgeType = contains AND type[aggregation])
    \_isFitted=false
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                88.495
  backend-query                                                     1001                      15.346
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
EdgeVertexStep(IN)                                                  1001        1001           7.951     0.25
HasStep([uniqueId.eq(interface831)])                                   1           1        2964.526    91.78
                                            >TOTAL                     -           -        3230.039        -
```
And now with the with the property pre-fetch optimisation:
```
gremlin> g.V(1050048).outE('aggregation').has('edgeType', 'contains').inV().has('uniqueId', 'interface831').profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
GraphStep(vertex,[1050048])                                            1           1          37.789     6.14
JanusGraphVertexStep([edgeType.eq(contains)])                       1001        1001         214.317    34.80
    \_condition=(edgeType = contains AND type[aggregation])
    \_isFitted=false
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                84.072
  backend-query                                                     1001                      14.929
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@81bcaf2b
JanusGraphEdgeVertexStep(IN)                                        1001        1001         247.477    40.19
    \_condition=(PROPERTY AND visibility:normal)
    \_isFitted=true
    \_vertices=1001
    \_multiPreFetch=true
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@8019d62e
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                 0.907
  backend-query                                                    13013                     216.843
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@8019d62e
HasStep([uniqueId.eq(interface831)])                                   1           1         116.257    18.88
                                            >TOTAL                     -           -         615.842        -
```

smcquillan commented:
Currently this optimisation would be automatically triggered if both of the following are true:
```
query.batch=true
query.fast-property=true
```
Was wondering if we ought to give more control by adding a new configuration option along the lines of:
```
query.batch-property-prefetch=true
```
or something along those lines.

FlorianHockmann commented:
Thanks for the explanations and for trying out the different traversals with your optimization.

> use of limit in a traversal like the following g.V().has('name', 'hercules').out().has('age', 5).limit(1) we've a chicken and egg situation as the limit is applied after the has step checks if the age is 5 and the optimisation needs to kick in before the has step so not sure there's much we can do there.

Wouldn't it be possible to check whether `has` is followed by `limit` and in that case not apply the optimization?
I really like the optimization you developed, but it leads to a performance hit for a traversal that uses `limit` in such a way which means that it should probably only be activated if users know that they won't use such traversals. If we can restrict the optimization to not kick in for such a traversal, then the optimization would really have no negative implications (at least I don't see any).

> Was wondering if we ought to give more control by adding a new configuration option

Good question. I'd say that it makes sense to add a new configuration option in either of these conditions:

1. Users won't expect such an optimization to kick in based on the docs of the existing options.
2. There are valid reasons for users to not want this new optimization if they have the other two options activated OR only to use this optimization, but not necessarily the other two.

If performance for traversals that use `limit()` like discussed above suffer from this optimization, then condition 2 warrants a new config option in my opinion.

smcquillan commented:
@FlorianHockmann I've modified things such that if there is a limit step following the has steps the optimisation will not kick in. And I've also added a new `query.batch-property-prefetch=true` configuration option allowing the new optimisation to be turned on or off an defaulted this to `false` so that existing users won't get any surprises.

FlorianHockmann commented on janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:
```diff
@@ -4023,12 +4021,36 @@ public void testTinkerPopOptimizationStrategies() {
+        gts = graph.traversal();
+
+        // This tests an edge property before inV and will trigger the multiQuery property pre-fetch optimisation in JanusGraphEdgeVertexStep
+        t = gts.V().has("id", sid).outE("knows").has("weight", P.between(1, 3)).inV().has("weight", P.between(1, 3)).profile("~metrics");
```
Can't you omit the `has` after the `inV` here as it's the `has` before the `inV` you want to test here?

smcquillan replied:
Actually we can't the `has` after the `inV` is the one that the optimisation is fetching the properties for. The first `has` is also needed as it's what sends us down the path which needs the `JanusGraphEdgeVertexStep`

FlorianHockmann commented on janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:
```diff
@@ -4023,12 +4021,36 @@ public void testTinkerPopOptimizationStrategies() {
+        // This tests an edge property before inV and will trigger the multiQuery property pre-fetch optimisation in JanusGraphEdgeVertexStep
+        t = gts.V().has("id", sid).outE("knows").has("weight", P.between(1, 3)).inV().has("weight", P.between(1, 3)).profile("~metrics");
+        assertNumStep(superV * (numV / 5 * 2), 2, (GraphTraversal)t, JanusGraphStep.class, JanusGraphVertexStep.class);
+        metrics = t.asAdmin().getSideEffects().get("~metrics");
```
(nitpick) How about moving this and the next line into a method like:
```java
assertMultiPrefetchUsed(Traversal t)
```

The test below that checks that it's not using multi prefetching could then use a method like `assertMultiPrefetchNotUsed`.

smcquillan replied:
Liked the idea of moving those lines into a method, but went for something a bit more generic so added `queryProfilerAnnotationIsPresent()` so I could also use it for testing the multiQuery annotation was being set as previously that wasn't being tested.

FlorianHockmann replied:
That's even better 👍

FlorianHockmann commented on janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:
```diff
@@ -4023,12 +4021,36 @@ public void testTinkerPopOptimizationStrategies() {
 
 
+        assertNumStep(superV * (numV / 5 * 2), 2, gts.V().has("id", sid).outE("knows").has("weight", P.between(1, 3)).inV().has("weight", P.between(1, 3)), JanusGraphStep.class, JanusGraphVertexStep.class);
+        assertNumStep(superV * (numV / 5 * 2), 2, gts.V().has("id", sid).outE("knows").inV().has("weight", P.between(1, 3)), JanusGraphStep.class, JanusGraphVertexStep.class);
```
I think you can remove this and the assert above now as you have included these now already below where it's also easier to understand them as you've added comments there.

smcquillan replied:
removed them

FlorianHockmann commented on janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:
```diff
@@ -4023,12 +4021,36 @@ public void testTinkerPopOptimizationStrategies() {
+        t = gts.V().has("id", sid).outE("knows").has("weight", P.between(1, 3)).inV().has("weight", P.between(1, 3)).profile("~metrics");
+        assertNumStep(superV * (numV / 5 * 2), 2, (GraphTraversal)t, JanusGraphStep.class, JanusGraphVertexStep.class);
+        metrics = t.asAdmin().getSideEffects().get("~metrics");
+        assertTrue(metrics.toString().contains(QueryProfiler.MULTIPREFETCH_ANNOTATION + "=true"));
```
Shouldn't we also test that this optimization won't be used if the `BATCH_PROPERTY_PREFETCHING` is not set?

smcquillan replied:
added an extra test for that

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphLocalQueryOptimizerStrategy.java:
```diff
@@ -92,8 +88,22 @@ public void apply(final Traversal.Admin<?, ?> traversal) {
+            }
 
-                preFetchPropertiesWithMultiQuery(janusGraph, vertexStep, traversal, nextStep);
+            if (janusGraph.getConfiguration().batchPropertyPrefetching()) {
```
good idea have moved it

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/query/vertex/MultiVertexCentricQueryBuilder.java:
```diff
@@ -140,6 +140,12 @@ public JanusGraphMultiVertexQuery addAllVertices(Collection<? extends Vertex> ve
 
+    @Override
+    public void preFetch() {
+        profiler.setAnnotation(QueryProfiler.MULTIPREFETCH_ANNOTATION,true);
```
done

smcquillan commented:
@FlorianHockmann suggested changes all looked good so have implemented them. Have also squashed everything into a single commit and rebased against master because, as you say, this is new functionality.

Incidentally do you know roughly when the next release might be?  I ask because for us to really benefit from multiQuery I think we'll need to address issue #1044 as well and if I've time I may look into what can be done there.

FlorianHockmann commented:
> Incidentally do you know roughly when the next release might be? 

The [v0.4.0 milestone](https://github.com/JanusGraph/janusgraph/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Release+v0.4.0%22) summarizes some issues that we want to fix for the 0.4.0 release. Most of them are just about dropping support for outdated backend versions which shouldn't require that much work. A bigger issue is the TinkerPop upgrade to 3.4.0 as that resulted in a lot of failed `GraphComputer` tests which probably requires some work on the `FulgoraGraphComputer`, an area I haven't worked on yet.
Then we also have a big PR currently open that changes the whole documentation generation process and which therefore requires a thorough review: #1437.
Therefore I'd say that some time in April sounds realistic for the 0.4.0 release.
So, you probably have enough time to tackle that issue.

smcquillan commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java:
```diff
@@ -1214,6 +1220,7 @@ public boolean apply(@Nullable String s) {
     private Boolean propertyPrefetching;
     private boolean adjustQueryLimit;
     private Boolean useMultiQuery;
+    private Boolean batchPropertyPrefetching;
```
Thanks for taking the time to review. And while I expect `query.batch-property-prefetch` probably could use `boolean` I followed the existing convention used by the `query.batch` (_useMultiQuery_) and `query.fast-property` (_propertyPrefetching_) both of which use the `Boolean`.

porunov replied:
@smcquillan I am OK here with both `Boolean` and `boolean` types.

FlorianHockmann commented:
@twilmes you seem to be the only one who has already worked on this part of the code base, so if you're available to review this in the next few weeks, then we can leave the PR open.
Otherwise, I'd say that we merge this with @porunov's and my approval.
